src/compiler/checker.ts

```diff
@@ -25164,87 +25138,30 @@ namespace ts {
             // the guard. For example: narrowing `{} | undefined` by `"boolean"` should produce the type `boolean`, not
             // the filtered type `{}`. For this reason we narrow constituents of the union individually, in addition to
             // filtering by type-facts.
-            function narrowTypeByImpliedType(type: Type, candidate: Type) {
-                if (type.flags & TypeFlags.AnyOrUnknown) {
-                    return candidate;
-                }
-                return mapType(type, t => {
-                    if (isTypeRelatedTo(t, candidate, strictSubtypeRelation)) {
-                        return t;
-                    }
-                    return mapType(candidate, c => {
-                        if (!areTypesComparable(t, c)) {
-                            return neverType;
-                        }
-                        if (c.flags & TypeFlags.Primitive && t.flags & TypeFlags.Object && !isEmptyAnonymousObjectType(t)) {
-                            return isTypeSubtypeOf(c, t) ? c : neverType;
-                        }
-                        if (c === globalFunctionType && isTypeSubtypeOf(c, t)) {
-                            return c;
-                        }
-                        return getIntersectionType([t, c]);
-                    });
-                });
+            function narrowTypeByTypeFacts(type: Type, impliedType: Type, facts: TypeFacts) {
+                return mapType(type, t =>

```

Comment:
Just a note from a bystander - I know that nested ternaries are pretty common in this codebase but they are making debugging & stepping through the code much harder.

Reply:
I hear you. But I much prefer the succinctness of the more functional style, and when you're looking at ternaries you sort of implicitly know that you don't have to worry about assignments and side effects.